### PR TITLE
Added Glyphs to accomidate color blind people.

### PR DESCRIPTION
**Pull request by ayourk**
Current pallet may make some glyphs difficult to see.
  Effort has been made to make these Islands of Insight copyright/trademark safe glyphs.
  https://jsfiddle.net/5gawy0np/20/ was used as a scratchpad.
  All glyphs scale according to their HTML5 canvas size (Y-axis) and were developed as 256px x 256px
  Glyphs require a Canvas and I do not yet know how to handle hover, so that is lost.
  See: https://stackoverflow.com/questions/29300280/update-html5-canvas-rectangle-on-hover
Allow toggling on/off of currently selected tool/color/glyph (selected glyph vs white).
  Selected glyph has to match existing tile to toggle.
Get color pallet from CSS (external style sheet); Won't work with "file:///" in Chrome. 
Fixed an HTML validation complaint for INPUT fields.

### CODE REVIEW

**Comment by MiniMage7** [comment]
This should really still be multiple pull requests. Adding glyphs is a whole feature by itself. Adding a textbox back to the website after it was removed in an early update is another update. Adding the deleted python file back to the project should probably be separate too. I don't really want to do a code review for 1300 lines across 5 files and multiple features.

**Comment by ayourk** [comment]
As I stated before, Best I can do is to separate out the different functionalities into different commits. Once a pull request is sent, any subsequent commits will show up too.

**Comment by MiniMage7** [comment]
Ah I missed that message. I'll probably have to wait until at least Thursday to look at it then. I don't have the time to code review this much at once.

**Comment by ayourk** [comment]
That's fine.  I realize I'm a prolific coder.

**Comment by MiniMage7** [comment]
First commit:
At a glance though, so you can look at some things:
In line 115 of styles, you left a comment saying you changed the color, but you didn't actually change the color.
You left some commented out code at line 486 of website.js
Start at line 562, half or your comments start lowercase (which isn't the biggest deal just a little weird)
1319 to end could probably be a for loop.

Second commit:
Huge commented out code at line 525 of website.js

And a question. why do you have `// reset composite mode to default` at the end of every draw glyph function without any code? Is there supposed to be something there? I'm not very knowledgeable about the drawing mechanics.

**Comment by ayourk** [comment]
I did a lot of copy paste from this example: https://jsfiddle.net/44mesdg8/ so it is just left over from that.

**Comment by ayourk** [comment]
See https://www.w3schools.com/graphics/canvas_intro.asp for HTML5 Canvas reference.

**Comment by ayourk** [comment]
> First commit: At a glance though, so you can look at some things: In line 115 of styles, you left a comment saying you changed the color, but you didn't actually change the color. 
I didn't change the color, but pointed to a value that looks better.  You stated you wanted to keep the stock colors you made.

>You left some commented out code at line 486 of website.js 
I was doing a lot of debugging for the CSS loading.

>1319 to end could probably be a for loop.
Granted, the glyph code could do with some optimization.

> Second commit: Huge commented out code at line 525 of website.js
This is documentation of example of how to use solution import stuff.  I don't know how else to put it.

**Comment by MiniMage7** [comment]
> This is documentation of example of how to use solution import stuff. I don't know how else to put it.

Honestly looking back at it, not sure why I thought it was code. Probably just say the brackets and went that's a function. It's fine there.

I'll try to talk to my visual design friend who made the color palette with me and see if I can get a color on theme that's a bit darker.

**Comment by ayourk** [comment]
> > This is documentation of example of how to use solution import stuff. I don't know how else to put it.
> 
> Honestly looking back at it, not sure why I thought it was code. Probably just say the brackets and went that's a function. It's fine there.
> 
> I'll try to talk to my visual design friend who made the color palette with me and see if I can get a color on theme that's a bit darker.

Please show my suggested color to your friend and see what he thinks.  See https://www.w3schools.com/colors/colors_picker.asp if you want an easy reference.

**Comment by MiniMage7** [comment]
I don't like the addition of a textbox at the bottom of the website. It doesn't seem to serve any real purpose while I was testing and makes the website look more cluttered. (After more testing it seems to have a use for importing things for Firefox, but I would rather not add a whole textbox to accommodate a single browser when its easy to import your board into another one.)
When exporting, after copying the export to the clipboard, the website also printed red text under the textbox?
When drawing with the color selector, going over the same tile twice deselects it even if its in the same brush stroke, which feels very off-putting.
I don't see a purpose to adding the solve JSONs to the website. The website is intended to be stand alone. Almost every user is only going to be using just the web part and that is intended. The only reason I added importing and exporting of boards is to share with other users, but the website itself will be able to do the solve itself. If your intention is to be able to communicate solves with the rust part of the program, then make the rust program output a pretty solution instead.

The adding of the glyphs to the tiles looks great, and if I could I would pull that in now, but you currently have all these changes under one branch so that isn't possible.

**Comment by ayourk** [comment]
> I don't like the addition of a textbox at the bottom of the website. It doesn't seem to serve any real purpose while I was testing and makes the website look more cluttered. (After more testing it seems to have a use for importing things for Firefox, but I would rather not add a whole textbox to accommodate a single browser when its easy to import your board into another one.) When exporting, after copying the export to the clipboard, the website also printed red text under the textbox? 
>When drawing with the color selector, going over the same tile twice deselects it even if its in the same brush stroke, which feels very off-putting. 

Since Hover can't be done with glyphs, being able to toggle the selected glyph on/off on the tile board makes quick work when a mistake is made.  I guess this could be disabled if still in "mouseDown".

>I don't see a purpose to adding the solve JSONs to the website. The website is intended to be stand alone. Almost every user is only going to be using just the web part and that is intended. 

Being able to import solution JSONs is to present a GUI for a play by play towards solution.  The plan was to post these solution JSONs to [Issue #6](/MiniMage7/MiniMage7.github.io/issues/6) as a reference.  I have 2 solves done and am working on the 3rd right now.  I was going to attach the solution JSON to the issue for others to use.  Each solution takes about 24-48 hours for me via Rust to complete.

>The only reason I added importing and exporting of boards is to share with other users, but the website itself will be able to do the solve itself. If your intention is to be able to communicate solves with the rust part of the program, then make the rust program output a pretty solution instead.

Then how do you propose that Rust shows a move by move solution?  Your text based move list is hard to follow.
> 
> The adding of the glyphs to the tiles looks great, and if I could I would pull that in now, but you currently have all these changes under one branch so that isn't possible.

**Comment by MiniMage7** [comment]
There are several ways that the rust display could be tackled, however I don't want any of them to be a part of the live website since it is a different application.

**Comment by MiniMage7** [comment]
Also, as I mentioned in https://github.com/MiniMage7/MiniMage7.github.io/issues/6, I'm planning on making a system that checks if the current solve is in the short list of impossible ones at the start of the solving process, and if it is, just return to the website that solution without going through the solve loop

**Comment by ayourk** [comment]
I understood the Rust application was to fill in the gaps for the weakness of the website.  Solution JSON allows that gap to be filled so that the solution doesn't need to be run through by EVERONE and just run once and shared to step through.

**Comment by MiniMage7** [comment]
> I understood the Rust application was to fill in the gaps for the weakness of the website. Solution JSON allows that gap to be filled so that the solution doesn't need to be run through by EVERONE and just run once and shared to step through.

This is a bit unrealistic since the number of people who are going to use the website, have an issue, and then check the issues and understand what they're reading is very small. According to google search console, less than 10 people click on website each day and I don't even show up on Bing.

**Comment by ayourk** [comment]
> Also, as I mentioned in [MiniMage7/MiniMage7.github.io#6](https://github.com/MiniMage7/MiniMage7.github.io/issues/6), I'm planning on making a system that checks if the current solve is in the short list of impossible ones at the start of the solving process, and if it is, just return to the website that solution without going through the solve loop

And this is where the Soltuion JSON can be stored in said database and then presented to the user.

**Comment by ayourk** [comment]
> > I understood the Rust application was to fill in the gaps for the weakness of the website. Solution JSON allows that gap to be filled so that the solution doesn't need to be run through by EVERONE and just run once and shared to step through.
> 
> This is a bit unrealistic since the number of people who are going to use the website, have an issue, and then check the issues and understand what they're reading is very small. According to google search console, less than 10 people click on website each day and I don't even show up on Bing.

My plan was to post a Steam Guide referencing your solver for these puzzles and listing the caveats.  That would get more traffic for you.

**Comment by MiniMage7** [comment]
If you're plan is to solve all of these yourself with rust and then compile them as a list, then how about instead just hard coding them after being solved instead of adding extra steps? It might be a bit awkward at first because of people potentially using different colors/a too large board size, but both of those could relatively easily be remedied, and that would make the experience way better for the end user
(For covering both of those issues. First one, checking if tiles are equal to each other rather than a constant. Second one, checking if a row/column is all less than 1, and then adding an offset if they are)

By doing the method this way, the website wouldn't get more cluttered, from an outside perspective and sharing a puzzle would be equivalent to sharing a solve.

**Comment by ayourk** [comment]
I believe the solvability of the puzzle can be calculated based on the width, height and number of tiles that have glyphs > 0.

**Comment by MiniMage7** [comment]
I more meant for comparing if its one of the solves that we have cached, but its not the number of tiles that makes a solve hard for the algorithm, its the numbers of possible moves that can be done at any given moment. The bot can actually succeed on huge puzzles if they have relatively linear solutions and fail on smaller ones that have a ton of options. In fact, the reason the checkForLoss() function was added was because the bot used to fail on something like this:
image

**Comment by ayourk** [comment]
For the puzzle you presented, it is 50 cells, with 50 active tiles, and 9 moves.

Another puzzle the web solver doesn't choke on:
{"width":9, "height":6, "board":[[0,0,0,1,2,3,4,2,0],[0,0,0,5,1,2,3,4,0],[0,0,0,3,4,5,1,6,0],[7,7,5,6,3,4,5,1,7],[-1,-1,-1,3,4,5,1,6,2],[-1,-1,-1,5,1,2,3,4,2]]}

This has 54 cells, 36 active tiles, and 10 moves.

Here are 3 puzzles that take too long for the website:
{ "width": 11, "height": 12, "moves":21, "swaps":[[[4,2], [4,3]], [[1,8], [1,9]], [[3,2], [3,3]], [[2,2], [3,2]], [[3,7], [3,8]], [[3,10], [4,10]], [[4,2], [4,3]], [[4,2], [4,3]], [[4,8], [4,9]], [[7,1], [7,2]], [[7,2], [7,3]], [[7,8], [7,9]], [[7,7], [7,8]], [[8,5], [9,5]], [[8,3], [8,4]], [[7,4], [8,4]], [[9,6], [10,6]], [[10,5], [11,5]], [[10,4], [11,4]], [[11,6], [11,7]], [[11,6], [11,7]]], "board": [[0,0,1,0,0,0,0,0,1,0,0], [0,1,7,7,0,0,0,0,2,1,8], [2,2,1,7,0,0,0,7,1,2,8], [2,2,8,2,0,0,0,2,7,8,2], [8,8,2,8,0,0,0,7,2,2,8], [-1,-1,6,-1,4,0,0,-1,6,-1,-1], [-1,-1,6,-1,4,3,0,-1,6,-1,-1], [1,6,1,1,6,8,0,1,1,6,1], [-1,-1,-1,5,5,3,3,5,-1,-1,-1], [-1,-1,-1,6,4,5,6,5,-1,-1,-1], [-1,-1,-1,6,4,6,5,6,-1,-1,-1], [-1,-1,-1,3,3,4,4,3,-1,-1,-1]] }

This has 132 cells, 69 active tiles, and 21 moves.

{"width":13, "height":14, "moves":13, "swaps":[[[2,10],[3,10]], [[2,0],[3,0]], [[6,4],[7,4]], [[2,2],[3,2]], [[5,6],[6,6]], [[6,4],[7,4]], [[7,10],[8,10]], [[7,8],[8,8]], [[7,12],[8,12]], [[8,0],[9,0]], [[8,2],[9,2]], [[9,4],[10,4]], [[10,6],[11,6]]], "board":[[0,0,1,0,5,0,0,0,1,0,5,0,6],[0,0,1,0,1,0,0,0,3,0,5,0,1],[7,0,5,0,1,0,0,0,3,0,1,0,7],[5,5,1,5,5,0,3,0,1,1,5,1,1],[7,-1,-1,-1,7,0,3,0,3,-1,-1,-1,7],[7,-1,-1,-1,5,0,6,0,3,-1,6,-1,7],[6,-1,3,-1,6,6,3,6,6,-1,6,-1,5],[6,-1,3,-1,5,-1,-1,-1,7,-1,3,-1,1],[1,-1,6,-1,6,-1,6,-1,3,3,6,3,3],[6,6,3,6,6,-1,6,-1,1,-1,-1,-1,5],[1,-1,-1,-1,1,-1,3,-1,1,-1,-1,-1,5],[1,-1,-1,-1,3,3,6,3,3,-1,-1,-1,1],[5,-1,-1,-1,7,-1,-1,-1,7,-1,-1,-1,6],[5,-1,-1,-1,7,-1,-1,-1,7,-1,-1,-1,6]]}

This has 182 cells, 78 active tiles, and 13 moves.

{ "width": 9, "height": 23, "board": [[0,0,0,0,0,0,0,4,0], [0,0,0,0,0,0,0,4,0], [0,0,0,0,0,0,0,1,0], [0,0,0,0,0,0,0,4,0], [0,0,0,0,0,0,0,5,0], [0,0,0,0,0,0,0,5,0], [0,0,0,0,6,0,0,2,0], [0,0,0,0,5,0,0,2,0], [0,0,0,0,1,0,0,6,0], [0,8,0,1,5,4,1,8,7], [7,2,5,4,6,5,4,2,5], [1,8,4,3,6,3,6,8,6], [1,8,4,1,3,4,1,8,7], [-1,6,-1,-1,2,-1,-1,6,-1], [-1,6,-1,-1,2,-1,-1,6,-1], [-1,3,-1,-1,8,-1,-1,3,-1], [-1,6,-1,-1,2,-1,-1,6,-1], [-1,3,-1,-1,8,-1,-1,3,-1], [-1,3,-1,-1,8,-1,-1,3,-1], [8,2,8,6,3,6,8,2,8], [2,8,2,3,6,3,2,8,2], [3,6,3,2,8,2,3,6,3], [6,3,6,8,2,8,6,3,6]] }

This has 207 cells, 100 active tiles, and I'm still calculating the number of moves.

If it really is moves driven, even if you could provide a partial solution (like say first 10 moves) that would be a great help if the solution isn't already known.  And if you provide partial solutions (10 moves at a time), you could repopulate the web solver every 10 moves to continue down the chain.

**Comment by MiniMage7** [comment]
I'm not sure you really understand how the algorithm works if you think it can provide the first 10 moves at a time or something. The only way it knows if a set of moves is correct is if it knows the whole path. If it knows the first 10 moves of a 14 move puzzle, then it can instantly figure out the last 4. If you try one of those impossible puzzles but do the first 2 moves or so, I wouldn't be surprised if the website could solve them instantly even with them having a large number of "active tiles".

I would test this myself but I'm currently at a doctors appointment.

**Comment by MiniMage7** [comment]
But that last puzzle for example, has 14 different moves the puzzle can start with, and depending on which move you choose, there can still be 14 moves possible for the second choice. This isn't an exact math, but the number of possible board states for this puzzle can be approximated by 14! or about 87 trillion. This is likely off by a bit, but even if its a factor of 10, its still 8 trillion.

If you knew the first 2 moves, it goes down to about 12! or 479 billion. So the puzzle become 100x easier.

Of course the number of tiles that can be moved does impact solve time, but that is way more outweighed by the number of board states.

**Comment by MiniMage7** [comment]
> But that last puzzle for example, has 14 different moves the puzzle can start with, and depending on which move you choose, there can still be 14 moves possible for the second choice. This isn't an exact math, but the number of possible board states for this puzzle can be approximated by 14! or about 87 trillion. This is likely off by a bit, but even if its a factor of 10, its still 8 trillion.
> 
> If you knew the first 2 moves, it goes down to about 12! or 479 billion. So the puzzle become 100x easier.
> 
> Of course the number of tiles that can be moved does impact solve time, but that is way more outweighed by the number of board states.

To further expand on this, this is actually how the rust implementation is optimizing solves. If you set the thread depth to 1, then what it's doing is (if there's 10 different first moves) making 10 threads that each start from one of those moves. If the depth is 2, and each of those have 10 more moves after, then it makes 100 threads with each possible start of 2 moves. If you had infinite CPU threads (and a decent chunk of ram) you could solve any puzzle near instantaneously.

**Comment by ayourk** [comment]
Thanks for that.  The problem I'm having is that if I increase the threads past 1, I get a core dump.  I'm pretty sure it boils down to proper error checking between each step somewhere.  I have yet to look into this and learn more Rust to handle all possible error conditions.

**Comment by MiniMage7** [comment]
> Thanks for that. The problem I'm having is that if I increase the threads past 1, I get a core dump. I'm pretty sure it boils down to proper error checking between each step somewhere. I have yet to look into this and learn more Rust to handle all possible error conditions.

 I'm pretty sure if you're getting a core dump, its just that your operating system is panicking that its making like ~215 threads that all demand a large amount of resources. Does it still provide the same error on more reasonably sized puzzles? I haven't had any crashing issues on windows until I get an error along the lines of there not being enough resources to create another thread.

**Comment by MiniMage7** [comment]
If the issue is just that it is too many threads for your OS, then I don't see any reason to change anything since that's the entire point of it asking how much depth you want to use at the start.

**Comment by ayourk** [comment]
```
thread 'thread '<unnamed><unnamed>' panicked at ' panicked at /build/rustc-QLyM7K/rustc-1.74.1+dfsg0ubuntu1~bpo0/library/std/src/thread/mod.rslibrary/std/src/sys/unix/stack_overflow.rs::686148::2913:
thread ':
failed to spawn thread: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }<unnamed>failed to set up alternative stack guard page: Cannot allocate memory (os error 12)
thread '' panicked at 
stack backtrace:
<unnamed>/build/rustc-QLyM7K/rustc-1.74.1+dfsg0ubuntu1~bpo0/library/std/src/thread/mod.rs' panicked at :/build/rustc-QLyM7K/rustc-1.74.1+dfsg0ubuntu1~bpo0/library/std/src/thread/mod.rs686::68629::
29failed to spawn thread: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }:
thread '
failed to spawn thread: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }<unnamed>' panicked at library/std/src/sys/unix/stack_overflow.rs:
148:13:
failed to set up alternative stack guard page: Cannot allocate memory (os error 12)
thread '<unnamed>' panicked at library/std/src/sys/unix/stack_overflow.rs:148:13:
failed to set up alternative stack guard page: Cannot allocate memory (os error 12)
thread 'thread '<unnamed><unnamed>' panicked at ' panicked at library/std/src/sys/unix/stack_overflow.rslibrary/std/src/sys/unix/stack_overflow.rs::144144::1313:
:
failed to allocate an alternative stack: Cannot allocate memory (os error 12)failed to allocate an alternative stack: Cannot allocate memory (os error 12)
```

**Comment by ayourk** [comment]
This article seems relevant to this current situation: https://stackoverflow.com/questions/167018/should-i-use-threading-and-recursion-together  One of the recommendations talks about using a thread queue with semaphores.

**Comment by MiniMage7** [comment]
Well, a lot of those answers aren't fully relevant to this. Usually when you're doing a task with multiple threads, it is a whole bunch of different tasks and they're usually waiting on other things to happen. You would this use something akin to semaphores to control data access etc.

But here we're simply using threads to try to the same task faster, which only works to the extent we have CPU resources available. The error you're getting is about what we would expect when you try to do more than your CPU can handle. "failed to spawn thread: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }"

Your operating system is basically saying "Hey I don't have strong enough hardware to make more threads here, so I'm going to stop whatever is trying to make the thread"

The way to fix this is to either get stronger hardware, or by lowering the thread depth.

**Comment by MiniMage7** [comment]
There might actually be a way to rewrite the program to use GPU for a lot of the repetitive processing which might fix several issues... but that's a bit outside of what I want to do with this project and I don't fully understand working with GPU's.

**Comment by ayourk** [comment]
> There might actually be a way to rewrite the program to use GPU for a lot of the repetitive processing which might fix several issues... but that's a bit outside of what I want to do with this project and I don't fully understand working with GPU's.

I do know that involves nVidia GPUs with CUDA or something like that.  I've looked into that but haven't gotten very far.

**Comment by ayourk** [comment]
Firefox clipboard reading requires setting `dom.events.asyncClipboard.readText` in `about:config` to `True`

**Comment by MiniMage7** [comment]
I found a few things mentioning that and similar, but I couldn't figure out how to implement it in practice or find anyone else's code showing it so I gave up. Very high chance I'm just being dumb, but if you figure it out, feel free to add it.

**Comment by MiniMage7** [comment]
I spent like 4 hours trying before giving up an adding the "Not available on Firefox" alert

**Comment by ayourk** [comment]
> I found a few things mentioning that and similar, but I couldn't figure out how to implement it in practice or find anyone else's code showing it so I gave up. Very high chance I'm just being dumb, but if you figure it out, feel free to add it.

The code you are using is correct; it is just the config flags that need to change to make it work.  You could put that in your error message to inform the user.
